rpc: update estimatesmartfee to return max of CBlockPolicyEstimator::estimateSmartFee, mempoolMinFee and minRelayTxFee

pranabp-bit opened the pull request:
This PR is in response to the issue [#19699](https://github.com/bitcoin/bitcoin/issues/19699).

Based on the discussion in the comments of PR [#22673](https://github.com/bitcoin/bitcoin/pull/22673) changes have been made in the `estimatesmartfee` itself such that it takes into account `mempoolMinFee` and `relayMinFee` . Hence it provides a fee estimate that is most likely to be paid by the user in an actual transaction, preventing issues such as [#16072](https://github.com/bitcoin/bitcoin/issues/16072).

The test file test/functional/feature_fee_estimation.py has also been updated to check this functionality.

pranabp-bit committed:
update estimatesmartfee rpc to return max of estimateSmartFee, mempoolMinFee and minRelayTxFee.

This will provide better estimates which would be closer to fee paid in actual
transactions.
The test has also been changed such that when the node is restarted with a
high mempoolMinFee, the estimatesmartfee still returns a feeRate greater
than or equal to the mempoolMinFee, minRelayTxFee.(just like the feeRate of actual transactions)

jonatack commented on src/rpc/mining.cpp:
```diff
@@ -1113,6 +1115,14 @@ static RPCHelpMan estimatesmartfee()
+    CFeeRate min_relay_feerate = ::minRelayTxFee;
+    if (feeRate < min_relay_feerate) {
+        feeRate = min_relay_feerate;
+    }
```
Suggestion, you could use braced initialization and `std::max` with an initializer list.
```diff
-    CFeeRate feeRate = fee_estimator.estimateSmartFee(conf_target, &feeCalc, conservative);
-    CFeeRate min_mempool_feerate = mempool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000);
-    if (feeRate < min_mempool_feerate) {
-        feeRate = min_mempool_feerate;
-    }
-    CFeeRate min_relay_feerate = ::minRelayTxFee;
-    if (feeRate < min_relay_feerate) {
-        feeRate = min_relay_feerate;
-    }
+    CFeeRate feeRate{fee_estimator.estimateSmartFee(conf_target, &feeCalc, conservative)};
+    CFeeRate min_mempool_feerate{mempool.GetMinFee(gArgs.GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000)};
+    CFeeRate min_relay_feerate{::minRelayTxFee};
+    feeRate = std::max({feeRate, min_mempool_feerate, min_relay_feerate});
```

jonatack commented on test/functional/feature_fee_estimation.py:
```diff
@@ -278,6 +282,11 @@ def run_test(self):
         check_estimates(self.nodes[1], self.fees_per_kb)
 
+        # check that the effective feerate is greater than or equal to the mempoolminfee even for high mempoolminfee
+        self.log.info("Estimates after restarting node with high MempoolMinFee")
```
```suggestion
        self.log.info("Test fee rate estimation after restarting node with high MempoolMinFee")
```

jonatack commented on test/functional/feature_fee_estimation.py:
```diff
@@ -133,9 +133,13 @@ def check_smart_estimates(node, fees_seen):
         feerate = float(e["feerate"])
         assert_greater_than(feerate, 0)
+        assert_greater_than_or_equal(feerate,float(mempoolMinFee))
+        assert_greater_than_or_equal(feerate,float(minRelaytxFee))
```
nit, missing space
```diff
-        assert_greater_than_or_equal(feerate,float(mempoolMinFee))
-        assert_greater_than_or_equal(feerate,float(minRelaytxFee))
+        assert_greater_than_or_equal(feerate, float(mempoolMinFee))
+        assert_greater_than_or_equal(feerate, float(minRelaytxFee))
```

jonatack reviewed (commented):
Concept ACK

jonatack commented on src/rpc/mining.cpp:
```diff
@@ -1097,6 +1097,8 @@ static RPCHelpMan estimatesmartfee()
     RPCTypeCheckArgument(request.params[0], UniValue::VNUM);
 
     CBlockPolicyEstimator& fee_estimator = EnsureAnyFeeEstimator(request.context);
+    NodeContext& node = EnsureAnyNodeContext(request.context);
```
can be `const`
```suggestion
    const NodeContext& node = EnsureAnyNodeContext(request.context);
```

prayank23 commented:
Concept ACK

`estimatesmartfee` will become better (smart) after this change

kristapsk commented:
Concept ACK

prayank23 commented:
Tried this on Pop!_OS with regtest. Initially had issues understanding how GetMinFee() works, Andrew Chow answered it in detail on stackexchange: https://bitcoin.stackexchange.com/questions/108126/getminfee-in-blockchain-cpp

I saved `minrelaytxfee` 0.0002 similar to [/test/functional/feature_fee_estimation.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_fee_estimation.py) and did some test transactions with different fee rate.

![image](https://user-images.githubusercontent.com/13405205/130057969-8b026c61-6c96-4b7b-ae24-f780c2b7fc9b.png)

`estimatesmartfee` suggests 20 sat/vB conf_target 2

Changed `minrelaytxfee` to 0.00001 (default value)

![image](https://user-images.githubusercontent.com/13405205/130073013-ad182d75-6a2a-4cf0-8356-8fd57ef5b132.png)

`estimatesmartfee` suggests 1 sat/vB conf_target 2

### Expected results:

```
$bitcoin-cli estimatesmartfee 1

{
  "feerate_min": 0.00001000
  "feerate_max": 0.00020000
  "blocks: 2
}
```
```
$bitcoin-cli estimatesmartfee 1

{
  "feerate_min": 0.00001000
  "feerate_max": 0.00100000
  "blocks: 2
}
```

`feerate_min` based on mempool fee rate distribution and `feerate_max` is the fee_rate returned right now 

_How to calculate `feerate_min` ?_

1. 2 * average_block_vsize where average_block_vsize is average virtual size of last N blocks

2. bytes from `getmempoolinfo`

3. if ((bytes/ 2 * average_block_vsize) < 1 use minimum fee rate from getmempoolinfo else use same as feerate_max.

_Why this will be useful?_

1. Users get a range instead of one fee rate 
2. Better estimates based on fee rates in mempool
3. Can use `feerate_min` for RBF transaction and replace with `feerate_max` if not confirmed in next block


`conf_target` works for 1 to 1008 blocks right now. This can be reduced to 1-10 if others agree to change it. 

<br><br>
<details>
  <summary> :bulb: One project which can be helpful in improving `estimatesmartfee`</summary>
  
  https://github.com/TrueLevelSA/btc-congestion-manager

  It returns mempool position which can be added in output for `estimatesmartfee`:

  
  ![image](https://user-images.githubusercontent.com/13405205/130071509-173b55f3-fa48-43d5-848c-f12340f562e8.png)


</details>

darosior reviewed (commented):
Concept ACK, but i think this should not be restricted to the `estimatesmartfee` RPC. Maybe we could have a wrapper to `CBlockPolicyEstimator`'s `estimateSmartFee` in `CTxMemPool` that would return the max of `minerPolicyEstimator->estimateSmartFee()` and `GetMinFee`? This way the node interface would call this wrapper instead of the estimator's one and it would apply this behaviour to all the current call sites.

pranabp-bit commented:
@prayank23 Thank you for the detailed review. Just wanted to confirm, are you suggesting to change `estimatesmartfee` such that is returns "feerate_min" and "feerate_max" ? If yes, then maybe I could try that later in a different PR.

pranabp-bit commented:
@darosior Thank you for the suggestion. But as of now, I only found one other call for the `estimateSmartFee` and it was in the [GetMinimumFeeRate function](https://github.com/bitcoin/bitcoin/blob/7fcf53f7b4524572d1d0c9a5fdc388e87eb02416/src/wallet/fees.cpp#L57). And this function returns the `feerate_needed` after considering the mempoolMinFee. So maybe we can keep this change limited to the `estimatesmartfee` RPC.

darosior commented:
> But as of now, I only found one other call

Sounds good to me, it's already a fix as is and we can always de-duplicate the `min(estimateSmarFee, mempoolMinFee)` logic in a folowup.

prayank23 commented:
> are you suggesting to change estimatesmartfee such that is returns "feerate_min" and "feerate_max" ? If yes, then maybe I could try that later in a different PR.

Yes. We could do something similar in follow up to improve `estimatesmartfee`. However, I also wanted to share that results for `estimatesmartfee` even after thing change are not helpful or misleading in some cases. 

For example: In first case mempool has some transactions with fee rate ranging from 1 to 100 sat/vB (less than 2MvB), estimatesmartfee should suggest fee rate close to 1 sat/vB for conf_target 2. It shows 20x what we actually need minimum at that moment.

instagibbs reviewed (approved):
ACK aside from test nit

pranabp-bit commented:
I updated the test such that in the end, it sets the `minrelaytxfee` to a value which is 3 times the `estimatesmartfee(1)` called from the same node. So the test would be more robust than the previous one in which I set it directly to "0.0003". 
I have incorporated most of the suggestions I got, and some of them, I would take up as follow-up in the future.
But for now, do suggest if any other change is required in the current PR.

laanwj commented:
Can you please change to a more descriptive PR title and commit message? "update estimatesmartfee" doesn't say much :smile:

instagibbs commented:
ACK https://github.com/pranabp-bit/bitcoin/pull/22722/commits/b2152f3ff21aeee185be6dcfcfcc73b28d989de2

meshcollider reviewed (commented):
utACK b2152f3ff21aeee185be6dcfcfcc73b28d989de2

darosior reviewed (commented):
utACK b2152f3ff21aeee185be6dcfcfcc73b28d989de2

meshcollider reviewed (commented):
re-utACK ea31caf6b4c182c6f10f136548f84e603800511c
